SopraGameController.cpp

```diff
@@ -1,5 +1,114 @@
+
+    template <>
+    double rng(double min, double max){
+        static std::random_device rd;
+        static std::default_random_engine el(rd());
+        std::uniform_real_distribution dist(min, max);
+        return dist(el);
+    }
+
+    template <>
+    int rng(int min, int max){
+        static std::random_device rd;
+        static std::default_random_engine el(rd());
+        std::uniform_int_distribution dist(min, max);
+        return dist(el);
+    }
+
+    bool actionTriggered(double actionProbability) {
+        if(actionProbability < 0 || actionProbability > 1){
+            throw std::invalid_argument("Probability not between 0 an 1");
+        }
+
+        return rng(0.0, 1.0) < actionProbability;
+    }
+
+    auto getAllCrossedCells(const gameModel::Position &startPoint, const gameModel::Position &endPoint) ->
+    std::vector<gameModel::Position> {
+
+        std::vector<gameModel::Position> resultVect;
+
+        // check if cells are valid
+        if (gameModel::Environment::getCell(startPoint) == gameModel::Cell::OutOfBounds ||
+        gameModel::Environment::getCell(endPoint) == gameModel::Cell::OutOfBounds){
+            throw std::out_of_range("Source or destination of movement vector are out of bounds");
+        }
+
+        // add the start point to the result
+        resultVect.emplace_back(startPoint);
+
+        // check if start and end point are equal
+        if (startPoint == endPoint)
+            return resultVect;
+
+        // define an normalize the direction vector
+        gameModel::Vector dirVect(endPoint.x - startPoint.x, endPoint.y - startPoint.y);
+        dirVect.normalize();
+
+        // traverse the route between the two points
+        gameModel::Vector travVect(0, 0);
+        gameModel::Position lastCell = startPoint;
+        while ((travVect + startPoint) != endPoint) {
+
+            // found a new crossed cell
+            if ((travVect + startPoint) != lastCell) {
+                lastCell = travVect + startPoint;
+                resultVect.emplace_back(lastCell);
+            }
+
+            // make a step to travers the vector
+            travVect = travVect + (dirVect * 0.5);
+        }
+
+        // add the end point to the result
+        resultVect.emplace_back(endPoint);
+
+        return resultVect;
+    }
+
+    auto getDistance(const gameModel::Position &startPoint, const gameModel::Position &endPoint) -> int {
+
+        // check if cells are valid
+        if (gameModel::Environment::getCell(startPoint) == gameModel::Cell::OutOfBounds ||
+            gameModel::Environment::getCell(endPoint) == gameModel::Cell::OutOfBounds){
+
+            return -1;
+        }
+
+        // check if start and end point are equal
+        if (startPoint == endPoint){
+            return 0;
+        }
+
+        int totalDistance = 0;
+
+        // calc the differences within the components of the given points
+        int dX = abs(startPoint.x - endPoint.x);
```

Comment:
```suggestion
        int dX = std::abs(startPoint.x - endPoint.x);
```
Wie bei pow abs ist C und nicht überladen, std::abs überladen für die meisten Typen